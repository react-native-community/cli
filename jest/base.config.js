/**
 * @flow
 */
module.exports = {
  rootDir: '../../',
  testEnvironment: 'node',
};
