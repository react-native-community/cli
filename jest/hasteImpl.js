/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow
 */

'use strict';

const path = require('path');
const findPlugins = require('../local-cli/core/findPlugins');

const REACT_NATIVE_CI = process.cwd() === path.resolve(__dirname, '..');

let pluginsPath;

if (REACT_NATIVE_CI) {
  pluginsPath = '..';
} else {
  pluginsPath = '../../../';
}

const plugins = findPlugins([path.resolve(__dirname, pluginsPath)]);

// Detect out-of-tree platforms and add them to the whitelists
const pluginRoots /*: Array<
  string,
> */ = plugins.haste.providesModuleNodeModules.map(
  name => path.resolve(__dirname, '../../', name) + path.sep,
);

const pluginNameReducers /*: Array<
  [RegExp, string],
> */ = plugins.haste.platforms.map(name => [
  new RegExp(`^(.*)\.(${name})$`),
  '$1',
]);

const ROOTS = [path.resolve(__dirname, '..') + path.sep, ...pluginRoots];

const BLACKLISTED_PATTERNS /*: Array<RegExp> */ = [
  /.*[\\\/]__(mocks|tests)__[\\\/].*/,
  /^Libraries[\\\/]Animated[\\\/]src[\\\/]polyfills[\\\/].*/,
  /^Libraries[\\\/]Renderer[\\\/]fb[\\\/].*/,
];

const WHITELISTED_PREFIXES /*: Array<string> */ = [
  'IntegrationTests',
  'Libraries',
  'ReactAndroid',
  'RNTester',
];

const NAME_REDUCERS /*: Array<[RegExp, string]> */ = [
  // extract basename
  [/^(?:.*[\\\/])?([a-zA-Z0-9$_.-]+)$/, '$1'],
  // strip .js/.js.flow suffix
  [/^(.*)\.js(\.flow)?$/, '$1'],
  // strip platform suffix
  [/^(.*)\.(android|ios|native)$/, '$1'],
  // strip plugin platform suffixes
  ...pluginNameReducers,
];

const haste = {
  /*
   * @return {string|void} hasteName for module at filePath; or undefined if
   *                       filePath is not a haste module
   */
  getHasteName(
    filePath /*: string */,
    sourceCode /*: ?string */,
  ) /*: string | void */ {
    if (!isHastePath(filePath)) {
      return undefined;
    }

    const hasteName = NAME_REDUCERS.reduce(
      (name, [pattern, replacement]) => name.replace(pattern, replacement),
      filePath,
    );

    return hasteName;
  },
};

function isHastePath(filePath /*: string */) /*: boolean */ {
  if (!filePath.endsWith('.js') && !filePath.endsWith('.js.flow')) {
    return false;
  }

  const root = ROOTS.find(r => filePath.startsWith(r));
  if (!root) {
    return false;
  }

  filePath = filePath.substr(root.length);
  if (BLACKLISTED_PATTERNS.some(pattern => pattern.test(filePath))) {
    return false;
  }
  return WHITELISTED_PREFIXES.some(prefix => filePath.startsWith(prefix));
}

module.exports = haste;
