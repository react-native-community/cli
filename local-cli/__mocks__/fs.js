/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const path = require('path');
const MemoryFS = require('metro-memory-fs');
let fs;

function setMockFilesystem(object, platform) {
  reset(platform);
  const root = platform === 'win32' ? 'c:\\' : '/';
  mockDir(root, {...object});
  return root;
}

function mockDir(dirPath, desc) {
  for (const entName in desc) {
    const ent = desc[entName];
    const entPath = path.join(dirPath, entName);
    if (typeof ent === 'string' || ent instanceof Buffer) {
      fs.writeFileSync(entPath, ent);
      continue;
    }
    if (typeof ent !== 'object') {
      throw new Error(require('util').format('invalid entity:', ent));
    }
    if (ent.SYMLINK != null) {
      fs.symlinkSync(ent.SYMLINK, entPath);
      continue;
    }
    fs.mkdirSync(entPath);
    mockDir(entPath, ent);
  }
}

function reset(platform) {
  if (path.mock == null) {
    throw new Error(
      'to use this "fs" module mock, you must also mock the "path" module',
    );
  }
  path.mock.reset(platform);
  const cwd = () => (platform === 'win32' ? 'c:\\' : '/');
  fs = new MemoryFS({platform, cwd});
  Object.assign(mockFs, fs);
}

const mockFs = {};
mockFs.__setMockFilesystem = setMockFilesystem;
mockFs.mock = {clear: reset};

reset('posix');

module.exports = mockFs;
