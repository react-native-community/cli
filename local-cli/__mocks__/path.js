/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const mockPath = {};

function reset(platform) {
  Object.assign(mockPath, jest.requireActual('path')[platform]);
}

mockPath.mock = {reset};

reset('posix');

module.exports = mockPath;
