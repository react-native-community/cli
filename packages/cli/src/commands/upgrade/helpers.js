// @flow
import https from 'https';

export const fetch = (url: string) =>
  new Promise<string>((resolve, reject) => {
    const request = https.get(url, response => {
      if (response.statusCode < 200 || response.statusCode > 299) {
        reject(
          new Error(`Failed to load page, status code: ${response.statusCode}`)
        );
      }
      const body = [];
      response.on('data', chunk => body.push(chunk));
      response.on('end', () => resolve(body.join('')));
    });
    request.on('error', err => reject(err));
  });
