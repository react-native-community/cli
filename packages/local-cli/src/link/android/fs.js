/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

const fs = require('fs-extra');

exports.readFile = file => () => fs.readFileSync(file, 'utf8');

exports.writeFile = (file, content) =>
  content
    ? fs.writeFileSync(file, content, 'utf8')
    : c => fs.writeFileSync(file, c, 'utf8');
