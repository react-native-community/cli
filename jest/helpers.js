// @flow
import fs from 'fs';
import os from 'os';
import path from 'path';
import {createDirectory} from 'jest-util';
import rimraf from 'rimraf';
import execa from 'execa';
import chalk from 'chalk';
import {Writable} from 'readable-stream';

const CLI_PATH = path.resolve(__dirname, '../packages/cli/build/bin.js');

type RunOptions = {
  nodeOptions?: string,
  nodePath?: string,
  timeout?: number, // kill the process after X milliseconds
  expectedFailure?: boolean,
};

export function run(
  dir: string,
  args?: Array<string>,
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  return spawnCli(dir, args, options);
}

// Runs cli until a given output is achieved, then kills it with `SIGTERM`
export async function runUntil(
  dir: string,
  args: Array<string> | void,
  text: string,
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  const spawnPromise = spawnCliAsync(dir, args, {timeout: 30000, ...options});

  spawnPromise.stderr.pipe(
    new Writable({
      write(chunk, _encoding, callback) {
        const output = chunk.toString('utf8');

        if (output.includes(text)) {
          spawnPromise.kill();
        }

        callback();
      },
    }),
  );

  return spawnPromise;
}

export const makeTemplate = (
  str: string,
): ((values?: Array<any>) => string) => (values?: Array<any>) =>
  str.replace(/\$(\d+)/g, (_match, number) => {
    if (!Array.isArray(values)) {
      throw new Error('Array of values must be passed to the template.');
    }
    return values[number - 1];
  });

export const cleanup = (directory: string) => rimraf.sync(directory);

/**
 * Creates a nested directory with files and their contents
 * writeFiles(
 *   '/home/tmp',
 *   {
 *     'package.json': '{}',
 *     'dir/file.js': 'module.exports = "x";',
 *   }
 * );
 */
export const writeFiles = (
  directory: string,
  files: {[filename: string]: string},
) => {
  createDirectory(directory);
  Object.keys(files).forEach(fileOrPath => {
    const dirname = path.dirname(fileOrPath);

    if (dirname !== '/') {
      createDirectory(path.join(directory, dirname));
    }
    fs.writeFileSync(
      path.resolve(directory, ...fileOrPath.split('/')),
      files[fileOrPath],
    );
  });
};

export const copyDir = (src: string, dest: string) => {
  const srcStat = fs.lstatSync(src);
  if (srcStat.isDirectory()) {
    if (!fs.existsSync(dest)) {
      fs.mkdirSync(dest);
    }
    fs.readdirSync(src).map(filePath =>
      copyDir(path.join(src, filePath), path.join(dest, filePath)),
    );
  } else {
    fs.writeFileSync(dest, fs.readFileSync(src));
  }
};

export const getTempDirectory = (name: string) =>
  path.resolve(os.tmpdir(), name);

function spawnCli(dir: string, args?: Array<string>, options: RunOptions = {}) {
  const {spawnArgs, spawnOptions} = getCliArguments({dir, args, options});

  const result = execa.sync(process.execPath, spawnArgs, spawnOptions);

  handleTestCliFailure(options, result, dir, args);

  return result;
}

function spawnCliAsync(
  dir: string,
  args?: Array<string>,
  options: RunOptions = {},
) {
  const {spawnArgs, spawnOptions} = getCliArguments({dir, args, options});

  try {
    return execa(process.execPath, spawnArgs, spawnOptions);
  } catch (result) {
    handleTestCliFailure(options, result, dir, args);
    return result;
  }
}

function getCliArguments({dir, args, options}) {
  const isRelative = !path.isAbsolute(dir);

  if (isRelative) {
    dir = path.resolve(__dirname, dir);
  }

  const env = Object.assign({}, process.env, {FORCE_COLOR: '0'});

  if (options.nodeOptions) {
    env.NODE_OPTIONS = options.nodeOptions;
  }
  if (options.nodePath) {
    env.NODE_PATH = options.nodePath;
  }

  const spawnArgs = [CLI_PATH, ...(args || [])];
  const spawnOptions = {
    cwd: dir,
    env,
    reject: false,
    timeout: options.timeout || 0,
  };
  return {spawnArgs, spawnOptions};
}

function handleTestCliFailure(options, result, dir, args) {
  if (!options.expectedFailure && result.code !== 0) {
    console.log(`Running CLI failed for unexpected reason. Here's more info:

${chalk.bold('dir:')}    ${dir}
${chalk.bold('args:')}   ${(args || []).join(' ')}
${chalk.bold('stderr:')} ${result.stderr}
${chalk.bold('stdout:')} ${result.stdout}
${chalk.bold('code:')}   ${result.code}`);
  }
}
