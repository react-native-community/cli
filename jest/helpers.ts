import fs from 'fs';
import os from 'os';
import path from 'path';
import {createDirectory} from 'jest-util';
// @ts-ignore jsfile
import rimraf from 'rimraf';
import execa from 'execa';
import chalk from 'chalk';
import slash from 'slash';
// @ts-ignore jsfile
import {Writable} from 'readable-stream';

const CLI_PATH = path.resolve(__dirname, '../packages/cli/build/bin.js');

type RunOptions = {
  nodeOptions?: string;
  nodePath?: string;
  timeout?: number; // kill the process after X milliseconds
  expectedFailure?: boolean;
};

/**
 * Helper function to run CLI command in a given folder
 */
export function runCLI(
  dir: string,
  args?: string[],
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  return spawnScript(process.execPath, [CLI_PATH, ...(args || [])], {
    ...options,
    cwd: dir,
  });
}

// Runs cli until a given output is achieved, then kills it with `SIGTERM`
export async function runUntil(
  dir: string,
  args: string[] | undefined,
  text: string,
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  const spawnPromise = spawnScriptAsync(dir, args || [], {
    timeout: 30000,
    cwd: dir,
    ...options,
  });

  spawnPromise.stderr.pipe(
    new Writable({
      write(chunk: any, _encoding: string, callback: () => void) {
        const output = chunk.toString('utf8');

        if (output.includes(text)) {
          spawnPromise.kill();
        }

        callback();
      },
    }),
  );

  return spawnPromise;
}

export const makeTemplate = (
  str: string,
): ((values?: Array<any>) => string) => (values?: Array<any>) =>
  str.replace(/\$(\d+)/g, (_match, number) => {
    if (!Array.isArray(values)) {
      throw new Error('Array of values must be passed to the template.');
    }
    return values[number - 1];
  });

export const cleanup = (directory: string) => rimraf.sync(directory);

/**
 * Creates a nested directory with files and their contents
 * writeFiles(
 *   '/home/tmp',
 *   {
 *     'package.json': '{}',
 *     'dir/file.js': 'module.exports = "x";',
 *   }
 * );
 */
export const writeFiles = (
  directory: string,
  files: {[filename: string]: string},
) => {
  createDirectory(directory);
  Object.keys(files).forEach(fileOrPath => {
    const dirname = path.dirname(fileOrPath);

    if (dirname !== '/') {
      createDirectory(path.join(directory, dirname));
    }
    fs.writeFileSync(
      path.resolve(directory, ...fileOrPath.split('/')),
      files[fileOrPath],
    );
  });
};

export const copyDir = (src: string, dest: string) => {
  const srcStat = fs.lstatSync(src);
  if (srcStat.isDirectory()) {
    if (!fs.existsSync(dest)) {
      fs.mkdirSync(dest);
    }
    fs.readdirSync(src).map(filePath =>
      copyDir(path.join(src, filePath), path.join(dest, filePath)),
    );
  } else {
    fs.writeFileSync(dest, fs.readFileSync(src));
  }
};

export const getTempDirectory = (name: string) =>
  path.resolve(os.tmpdir(), name);

type SpawnOptions = RunOptions & {
  cwd: string;
};

type SpawnFunction<T> = (
  execPath: string,
  args: string[],
  options: SpawnOptions,
) => T;

export const spawnScript: SpawnFunction<execa.ExecaReturns> = (
  execPath,
  args,
  options,
) => {
  const result = execa.sync(execPath, args, getExecaOptions(options));

  handleTestFailure(execPath, options, result, args);

  return result;
};

const spawnScriptAsync: SpawnFunction<execa.ExecaChildProcess> = (
  execPath,
  args,
  options,
) => {
  try {
    return execa(execPath, args, getExecaOptions(options));
  } catch (result) {
    handleTestFailure(execPath, options, result, args);
    return result;
  }
};

function getExecaOptions(options: SpawnOptions) {
  const isRelative = !path.isAbsolute(options.cwd);

  const cwd = isRelative ? path.resolve(__dirname, options.cwd) : options.cwd;

  const env = Object.assign({}, process.env, {FORCE_COLOR: '0'});

  if (options.nodeOptions) {
    env.NODE_OPTIONS = options.nodeOptions;
  }
  if (options.nodePath) {
    env.NODE_PATH = options.nodePath;
  }

  return {
    cwd,
    env,
    reject: false,
    timeout: options.timeout || 0,
  };
}

function handleTestFailure(
  cmd: string,
  options: SpawnOptions,
  result: {[key: string]: any},
  args: string[] | undefined,
) {
  if (!options.expectedFailure && result.code !== 0) {
    console.log(`Running ${cmd} command failed for unexpected reason. Here's more info:
${chalk.bold('cmd:')}     ${cmd}    
${chalk.bold('options:')} ${JSON.stringify(options)}
${chalk.bold('args:')}    ${(args || []).join(' ')}
${chalk.bold('stderr:')}  ${result.stderr}
${chalk.bold('stdout:')}  ${result.stdout}
${chalk.bold('code:')}    ${result.code}`);
  }
}

export function replaceProjectRootInOutput(output: string, testFolder: string) {
  const regex = new RegExp(`(:\\s").*(${slash(testFolder)})`, 'g');
  return slash(output).replace(regex, '$1<<REPLACED_ROOT>>');
}
