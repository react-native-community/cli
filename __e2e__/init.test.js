// @flow
import fs from 'fs';
import path from 'path';
import {run, getTempDirectory, cleanup, writeFiles} from '../jest/helpers';

const DIR = getTempDirectory('command-init');

beforeEach(() => {
  cleanup(DIR);
  writeFiles(DIR, {});
});
afterEach(() => {
  cleanup(DIR);
});

test('init --template fails without package name', () => {
  const {stderr} = run(
    DIR,
    ['init', '--template', 'react-native-new-template'],
    {expectedFailure: true},
  );
  expect(stderr).toContain('missing required argument');
});

test('init --template', () => {
  const templateFiles = [
    '.buckconfig',
    '.eslintrc.js',
    '.flowconfig',
    '.gitattributes',
    // should be here, but it's not published yet
    // '.gitignore',
    '.watchmanconfig',
    'App.js',
    '__tests__',
    'android',
    'babel.config.js',
    'index.js',
    'ios',
    'metro.config.js',
    'node_modules',
    'package.json',
    'yarn.lock',
  ];

  const {stdout} = run(DIR, [
    'init',
    '--template',
    'react-native-new-template',
    'TestInit',
  ]);

  expect(stdout).toContain('Welcome to React Native!');
  expect(stdout).toContain('Run instructions');

  // make sure we don't leave garbage
  expect(fs.readdirSync(DIR)).toEqual(['TestInit']);
  expect(fs.readdirSync(path.join(DIR, 'TestInit'))).toEqual(templateFiles);

  const pkgJson = require(path.join(DIR, 'TestInit', 'package.json'));
  expect(pkgJson).toMatchSnapshot(
    'package.json contains necessary configuration',
  );
});

test('init --template file:/tmp/custom/template', () => {
  writeFiles(DIR, {
    'custom/template/template.config.js': `module.exports = {
      placeholderName: 'HelloWorld',
      templateDir: './template-dir',
    };`,
    'custom/template/package.json':
      '{ "name": "template", "version": "0.0.1" }',
    'custom/template/template-dir/package.json': '{}',
    'custom/template/template-dir/empty': '',
  });

  const {stdout} = run(DIR, [
    'init',
    '--template',
    `file://${path.resolve(DIR, 'custom', 'template')}`,
    'TestInit',
  ]);

  expect(stdout).toContain('Run instructions');
});
