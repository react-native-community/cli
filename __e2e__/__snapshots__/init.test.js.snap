// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init --template: package.json contains necessary configuration 1`] = `
Object {
  "dependencies": Object {
    "react": "16.8.1",
    "react-native": "0.59.0-rc.1",
  },
  "devDependencies": Object {
    "@babel/core": "^7.3.3",
    "@babel/runtime": "^7.3.1",
    "@react-native-community/eslint-config": "^0.0.3",
    "babel-jest": "^24.1.0",
    "jest": "^24.1.0",
    "metro-react-native-babel-preset": "^0.51.1",
    "react-test-renderer": "16.8.1",
  },
  "jest": Object {
    "preset": "react-native",
  },
  "name": "TestInit",
  "private": true,
  "scripts": Object {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest",
  },
  "version": "0.0.1",
}
`;
